src/Functions/FunctionsComparison.h

```diff
@@ -1232,11 +1235,20 @@ class FunctionComparison : public IFunction
             }
             else
             {
-                // compare
+                /// Check does another data type is comparable to Decimal, includes Int and Float.
                 if (!allowDecimalComparison(left_type, right_type) && !date_and_datetime)
                     throw Exception(
                         "No operation " + getName() + " between " + left_type->getName() + " and " + right_type->getName(),
                         ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+                /// When Decimal comparing to Float32/64, we convert both of them into Float64.
+                /// Other systems like MySQL and Spark also do as this.
+                if (left_is_float || right_is_float)
+                {
+                    const auto converted_type = std::make_shared<DataTypeFloat64>();
+                    ColumnPtr c0_converted = castColumn(col_with_type_and_name_left, converted_type);
+                    ColumnPtr c1_converted = castColumn(col_with_type_and_name_right, converted_type);
+                    return executeGenericIdenticalTypes(c0_converted.get(), c1_converted.get());
```

Comment:
Maybe use more efficient implementation than generic? Something like `executeNum...`. Or for simplicity you can call `executeImpl` once again with new arguments.